plugin/core/rpc.py

```diff
@@ -69,11 +38,76 @@ def try_terminate_process(process: 'subprocess.Popen') -> None:
+        self.server_name = server_name
+        self.sink = sink
+
+    def log(self, message: str, params: 'Any', log_payload: bool) -> None:
+        if log_payload:
+            message = "{}: {}".format(message, params)
+        self.sink(message)
+
+    def format_response(self, direction: str, request_id: int) -> str:
+        return "{} {} {}".format(direction, self.server_name, request_id)
+
+    def format_request(self, direction: str, method: str, request_id: int) -> str:
+        return "{} {} {}({})".format(direction, self.server_name, method, request_id)
+
+    def format_notification(self, direction: str, method: str) -> str:
+        return "{} {} {}".format(direction, self.server_name, method)
+
+    def outgoing_response(self, request_id: int, params: 'Any') -> None:
+        if not self.settings.log_debug:
+            return
+        self.log(self.format_response(Direction.Outgoing, request_id), params, self.settings.log_payloads)
+
+    def outgoing_request(self, request_id: int, method: str, params: 'Any', blocking: bool) -> None:
+        if not self.settings.log_debug:
+            return
+        direction = Direction.OutgoingBlocking if blocking else Direction.Outgoing
+        self.log(self.format_request(direction, method, request_id), params, self.settings.log_payloads)
+
+    def outgoing_notification(self, method: str, params: 'Any') -> None:
+        if not self.settings.log_debug:
+            return
+        log_payload = self.settings.log_payloads \
+            and method != "textDocument/didChange" \
+            and method != "textDocument/didOpen"
+        self.log(self.format_notification(Direction.Outgoing, method), params, log_payload)
+
+    def incoming_response(self, request_id: int, params: 'Any') -> None:
+        if not self.settings.log_debug:
+            return
+        self.log(self.format_response(Direction.Incoming, request_id), params, self.settings.log_payloads)
+
+    def incoming_request(self, request_id: int, method: str, params: 'Any', unhandled: bool) -> None:
+        if not self.settings.log_debug:
+            return
+        direction = "unhandled" if unhandled else Direction.Incoming
+        self.log(self.format_request(direction, method, request_id), params, self.settings.log_payloads)
+
+    def incoming_notification(self, method: str, params: 'Any', unhandled: bool) -> None:
+        if not self.settings.log_debug or method == "window/logMessage":
+            return
+        direction = "unhandled" if unhandled else Direction.Incoming
+        self.log(self.format_notification(direction, method), params, self.settings.log_payloads)
+
+
 class Client(object):
     def __init__(self, transport: Transport, settings: Settings) -> None:
         self.transport = transport  # type: Optional[Transport]
         self.transport.start(self.receive_payload, self.on_transport_closed)
         self.request_id = 0
+        self.logger = PreformattedPayloadLogger(settings, "server", debug)
```

Comment:
A client will print to the ST console by default. The sink is modified in WindowManager to print to panel instead.